inst/shiny/functions/apply_column_mapping.R

```diff
@@ -0,0 +1,93 @@
+#' Apply UI-Based Column Mapping to a Dataset
+#'
+#' This function takes a dataset and applies user-specified column mappings
+#'  provided through a Shiny `input` object. It renames columns based
+#'  on the selected mappings, handles special cases such as `ADOSEDUR`,
+#'  updates units for key variables, applies labels,
+#'  and removes detected duplicate concentration records.
+#'
+#' @param dataset A data frame containing the raw input data.
+#' @param input A Shiny input object containing the selected column mappings
+#'  from the user interface. Expected to include input IDs like `select_<column_name>`.
+#' @param manual_units A list with named elements (`concentration`, `dose`, and `time`)
+#'  specifying the manually selectable unit options for relevant columns.
+#'
+#' @returns A transformed data frame with:
+#'   - Renamed columns according to user mappings
+#'   - Unit columns updated (if user-specified units are in `manual_units`)
+#'   - Concentration duplicates removed
+#'   - Labels applied to columns (via `apply_labels()`)
+#'
+#' @details
+#' - Logs the mapping selections for debugging.
+#' - Validates that all required columns are mapped and no duplicates exist.
+#' - If `ADOSEDUR` is mapped to `"NA"`, it is assigned a value of `0`.
+#' - Removes concentration data duplicates using all columns except `ARRLT`, `NRRLT`, and `DOSNO`.
+#' - Uses global objects like `MAPPING_COLUMN_GROUPS`, `MAPPING_DESIRED_ORDER`, and `LABELS`.
+
+apply_column_mapping <- function(dataset, mapping, manual_units, column_groups, desired_order) {
+  selected_cols <- sapply(names(column_groups), function(group) {
+    sapply(column_groups[[group]], function(column) {
+      mapping[[paste0("select_", column)]]
+    })
+  }, simplify = FALSE)
+
+  mapping_log <- lapply(selected_cols, function(group) {
+    purrr::imap(group, function(v, n) paste0("* ", n, " -> ", paste0(v, collapse = ", ")))
+  }) %>%
+    unlist() %>%
+    paste0(collapse = "\n")
+  log_debug(paste0("The following mapping was applied:\n", mapping_log))
+
+  if (any(unlist(selected_cols) == "")) {
+    log_error("Unmapped columns detected.")
+    showNotification("Some required columns are not mapped.
+                     Please complete all selections.", type = "error", duration = 5)
+    return()
+  }
+
+  if (any(duplicated(unlist(selected_cols)))) {
+    log_error("Duplicate column selection detected.")
+    showNotification("Duplicate column selection detected.
+                     Please ensure each selection is unique.", type = "error", duration = 5)
+    return()
```

Comment:
nitpick: Same as before
```suggestion
    return(NULL)
```